File: server/src/modules/pricing.js
import { z } from 'zod';
import { round, normalize, weightedScore, band } from '../utils/scoring.js';

/* ============================================================
   MODULE 5 — PRICING
   Competitive pricing analysis: list the competitors, surface
   their pricing models, and position our price against them.
   ============================================================ */

export const key = 'pricing';
export const title = 'Pricing';
export const action = 'MAPPING';

export const PRICING_MODELS = [
  'subscription',
  'usage-based',
  'per-seat',
  'one-time',
  'freemium',
  'commission',
  'tiered',
  'custom',
];

const competitorSchema = z.object({
  name: z.string().min(1),
  price: z.number().nonnegative(),
  model: z.enum(PRICING_MODELS).default('subscription'),
  billingPeriod: z.enum(['monthly', 'annual', 'one-time']).default('monthly'),
  notes: z.string().optional(),
  features: z.array(z.string()).default([]),
});

export const inputSchema = z.object({
  ourPrice: z.number().nonnegative(),
  ourModel: z.enum(PRICING_MODELS).default('subscription'),
  ourFeatures: z.array(z.string()).default([]),
  currency: z.string().default('USD'),
  competitors: z.array(competitorSchema).min(1, 'Supply at least one competitor to compare against'),
});

/** Normalise every price to a monthly figure so the comparison is like-for-like. */
function toMonthly(price, billingPeriod) {
  if (billingPeriod === 'annual') return price / 12;
  if (billingPeriod === 'one-time') return price / 36; // amortised over 3 years
  return price;
}

export function run(input) {
  const data = inputSchema.parse(input);

  const normalised = data.competitors.map((c) => ({
    ...c,
    monthlyPrice: round(toMonthly(c.price, c.billingPeriod), 2),
  }));

  const prices = normalised.map((c) => c.monthlyPrice).sort((a, b) => a - b);
  const median = prices.length % 2
    ? prices[(prices.length - 1) / 2]
    : round((prices[prices.length / 2 - 1] + prices[prices.length / 2]) / 2, 2);
  const min = prices[0];
  const max = prices[prices.length - 1];
  const mean = round(prices.reduce((a, b) => a + b, 0) / prices.length, 2);

  const delta = median ? round(((data.ourPrice - median) / median) * 100, 1) : 0;
  const position = delta <= -15 ? 'BELOW_MARKET' : delta >= 15 ? 'ABOVE_MARKET' : 'AT_MARKET';

  // Differentiation: which of our features nobody else lists.
  const competitorFeatures = new Set(normalised.flatMap((c) => c.features.map((f) => f.toLowerCase())));
  const uniqueFeatures = data.ourFeatures.filter((f) => !competitorFeatures.has(f.toLowerCase()));
  const parityFeatures = data.ourFeatures.filter((f) => competitorFeatures.has(f.toLowerCase()));

  // Pricing power comes from differentiation, not from being cheapest.
  const differentiationScore = normalize(uniqueFeatures.length, 0, 4);
  const positionScore = position === 'AT_MARKET' ? 80 : position === 'ABOVE_MARKET' ? 65 : 55;
  const modelDiversityScore = normalize(new Set(normalised.map((c) => c.model)).size, 1, 4);

  const score = weightedScore([
    { value: differentiationScore, weight: 0.45 },
    { value: positionScore, weight: 0.35 },
    { value: modelDiversityScore, weight: 0.2 },
  ]);

  return {
    score,
    output: {
      currency: data.currency,
      ourPrice: data.ourPrice,
      ourModel: data.ourModel,
      market: { min, max, median, mean, sampleSize: prices.length },
      position,
      deltaFromMedianPercent: delta,
      competitors: normalised
        .map((c) => ({
          name: c.name,
          price: c.price,
          monthlyPrice: c.monthlyPrice,
          model: c.model,
          billingPeriod: c.billingPeriod,
          cheaperThanUs: c.monthlyPrice < data.ourPrice,
          notes: c.notes ?? null,
        }))
        .sort((a, b) => a.monthlyPrice - b.monthlyPrice),
      modelsInMarket: [...new Set(normalised.map((c) => c.model))],
      differentiation: {
        unique: uniqueFeatures,
        parity: parityFeatures,
        uniqueCount: uniqueFeatures.length,
      },
      pricingPower: band(score),
      summary:
        `Priced ${Math.abs(delta)}% ${delta >= 0 ? 'above' : 'below'} the ${data.currency} ${median} market median ` +
        `across ${prices.length} competitor(s); ${uniqueFeatures.length} differentiating feature(s).`,
      recommendations: buildRecommendations({ position, delta, uniqueFeatures, data }),
    },
  };
}

function buildRecommendations({ position, delta, uniqueFeatures, data }) {
  const out = [];
  if (position === 'ABOVE_MARKET' && uniqueFeatures.length === 0) {
    out.push(`Priced ${delta}% above median with no differentiating features — justify the premium or reprice.`);
  }
  if (position === 'BELOW_MARKET' && uniqueFeatures.length >= 2) {
    out.push('Underpriced relative to differentiation — there is room to raise price.');
  }
  if (data.competitors.length < 3) out.push('Add more competitors: a sample of fewer than 3 makes the median unreliable.');
  if (!out.length) out.push('Pricing is defensible against the current competitive set.');
  return out;
}

export default { key, title, action, inputSchema, run };

File: server_python/modules/pricing.py
"""MODULE 5 — PRICING (port of server/src/modules/pricing.js)
Competitive pricing analysis: list the competitors, surface their pricing
models, and position our price against them.
"""
from typing import List, Literal, Optional
from pydantic import BaseModel, Field

from scoring import js_round, normalize, weighted_score, band

KEY = 'pricing'
TITLE = 'Pricing'
ACTION = 'MAPPING'

PRICING_MODELS = [
    'subscription', 'usage-based', 'per-seat', 'one-time',
    'freemium', 'commission', 'tiered', 'custom',
]

PricingModel = Literal['subscription', 'usage-based', 'per-seat', 'one-time',
                       'freemium', 'commission', 'tiered', 'custom']
BillingPeriod = Literal['monthly', 'annual', 'one-time']


class Competitor(BaseModel):
    name: str = Field(min_length=1)
    price: float = Field(ge=0)
    model: PricingModel = 'subscription'
    billingPeriod: BillingPeriod = 'monthly'
    notes: Optional[str] = None
    features: List[str] = []


class InputSchema(BaseModel):
    ourPrice: float = Field(ge=0)
    ourModel: PricingModel = 'subscription'
    ourFeatures: List[str] = []
    currency: str = 'USD'
    competitors: List[Competitor] = Field(min_length=1, description='Supply at least one competitor to compare against')


def _to_monthly(price, billing_period):
    """Normalise every price to a monthly figure so the comparison is like-for-like."""
    if billing_period == 'annual':
        return price / 12
    if billing_period == 'one-time':
        return price / 36  # amortised over 3 years
    return price


def run(raw_input, context=None):
    data = InputSchema.model_validate(raw_input or {})

    normalised = [
        {**c.model_dump(), 'monthlyPrice': js_round(_to_monthly(c.price, c.billingPeriod), 2)}
        for c in data.competitors
    ]

    prices = sorted(c['monthlyPrice'] for c in normalised)
    n = len(prices)
    median = prices[(n - 1) // 2] if n % 2 else js_round((prices[n // 2 - 1] + prices[n // 2]) / 2, 2)
    mean = js_round(sum(prices) / n, 2)

    delta = js_round(((data.ourPrice - median) / median) * 100, 1) if median else 0
    position = 'BELOW_MARKET' if delta <= -15 else 'ABOVE_MARKET' if delta >= 15 else 'AT_MARKET'

    # Differentiation: which of our features nobody else lists.
    competitor_features = {f.lower() for c in normalised for f in c['features']}
    unique_features = [f for f in data.ourFeatures if f.lower() not in competitor_features]
    parity_features = [f for f in data.ourFeatures if f.lower() in competitor_features]

    # Pricing power comes from differentiation, not from being cheapest.
    differentiation_score = normalize(len(unique_features), 0, 4)
    position_score = 80 if position == 'AT_MARKET' else 65 if position == 'ABOVE_MARKET' else 55
    model_diversity_score = normalize(len({c['model'] for c in normalised}), 1, 4)

    score = weighted_score([
        {'value': differentiation_score, 'weight': 0.45},
        {'value': position_score, 'weight': 0.35},
        {'value': model_diversity_score, 'weight': 0.2},
    ])

    return {
        'score': score,
        'output': {
            'currency': data.currency,
            'ourPrice': data.ourPrice,
            'ourModel': data.ourModel,
            'market': {'min': prices[0], 'max': prices[-1], 'median': median, 'mean': mean, 'sampleSize': n},
            'position': position,
            'deltaFromMedianPercent': delta,
            'competitors': sorted(
                (
                    {
                        'name': c['name'],
                        'price': c['price'],
                        'monthlyPrice': c['monthlyPrice'],
                        'model': c['model'],
                        'billingPeriod': c['billingPeriod'],
                        'cheaperThanUs': c['monthlyPrice'] < data.ourPrice,
                        'notes': c['notes'],
                    }
                    for c in normalised
                ),
                key=lambda c: c['monthlyPrice'],
            ),
            'modelsInMarket': list(dict.fromkeys(c['model'] for c in normalised)),
            'differentiation': {
                'unique': unique_features,
                'parity': parity_features,
                'uniqueCount': len(unique_features),
            },
            'pricingPower': band(score),
            'summary':
                f'Priced {abs(delta)}% {"above" if delta >= 0 else "below"} the {data.currency} {median} '
                f'market median across {n} competitor(s); {len(unique_features)} differentiating feature(s).',
            'recommendations': _build_recommendations(position, delta, unique_features, data),
        },
    }


def _build_recommendations(position, delta, unique_features, data):
    out = []
    if position == 'ABOVE_MARKET' and len(unique_features) == 0:
        out.append(f'Priced {delta}% above median with no differentiating features — justify the premium or reprice.')
    if position == 'BELOW_MARKET' and len(unique_features) >= 2:
        out.append('Underpriced relative to differentiation — there is room to raise price.')
    if len(data.competitors) < 3:
        out.append('Add more competitors: a sample of fewer than 3 makes the median unreliable.')
    if not out:
        out.append('Pricing is defensible against the current competitive set.')
    return out
